### DISCUSSION TITLE
How to toggle channel ON/OFF

### DISCUSSION THREAD

**Discussion by JackKelly1** [Telecominfraproject/oopt-gnpy#389] [Q&A] [answered]
I am currently trying to simulate a 40 channel link and want the ability to turn certain channels off. I have tried changing sim_params._nli_params._computed_channels but this only changes the calculations when using raman fibers. Please could you tell me how to toggle the channels when using SSMF for example?

Thank you.

**Comment by jktjkt** [comment, marked as answer]
This is currently not possible via the standard frontend (XLS and JSON files), only by manipulating the Python code directly. We're working on making it fully supported, along with non-uniform modulation rates.

**Reply by JackKelly1** [reply]
Thank you for the response @jktjkt. Please could you describe where in cli_examples.py (I assume) the changes should be made?

**Reply by jktjkt** [reply]
If you're comfortable with Python, take a look at the [patch series](https://review.gerrithub.io/q/project:Telecominfraproject/oopt-gnpy+topic:mixed-rate), but beware, it's a bit more involved. By all means please provide feedback.

For a quick-n-dirty approach, you might try [overriding the `create_input_spectral_information` function](https://github.com/Telecominfraproject/oopt-gnpy/blob/f98eb2c/gnpy/core/info.py#L45-L56) which is where we take the user's input about the uniform grid. Just skipping a channel or two should be reasonably straightforward. Here's an untested patch:
```diff
diff --git a/gnpy/core/info.py b/gnpy/core/info.py
index 1d0ac51..db288fb 100644
--- a/gnpy/core/info.py
+++ b/gnpy/core/info.py
@@ -50,7 +50,7 @@ def create_input_spectral_information(f_min, f_max, roll_off, baud_rate, power,
         pref=Pref(pref, pref, lin2db(nb_channel)),
         carriers=[
             Channel(f, (f_min + spacing * f),
-                    baud_rate, roll_off, Power(power, 0, 0), 0, 0) for f in range(1, nb_channel + 1)
+                    baud_rate, roll_off, Power(power, 0, 0), 0, 0) for f in range(1, nb_channel + 1) if f != 10
         ]
     )
     return si
```

**Reply by caffery-chen** [reply]
Is it possible to simulate non-uniform modulation rates in the latest release by manipulating e.g. the bandwidth when creating spectrum via SI ?

**Reply by jktjkt** [reply]
My understanding is that the results won't be accurate. There's a new version of the simulation engine that I pointed to where this error is addressed.  It's currently a work-in-progress, people are working in integrating that library with the rest of the GNPy. We're targeting a release with that for some time around OFC, but no guarantees on that one.
